### PR TITLE
Authorities refactoring

### PR DESCRIPTION
**Pull request by oldalton**

### CODE REVIEW

File: MSAL/src/instance/MSALOauth2ProviderFactory.m [hunk 1]
```diff
@@ -25,34 +25,44 @@
...
     }
-
-    // Create AAD v2 factory for everything else, but in future we might want to further separate this out
+    else if ([authority isKindOfClass:[MSALAADAuthority class]])
+    {
+        return [MSALAADOauth2Provider new];
+    }
+    else if ([authority isKindOfClass:[MSALADFSAuthority class]])
+    {
+        NSAssert(NO, @"ADFS not implemented in MSAL yet");
```
Comment (unpluggedk) [inline MSAL/src/instance/MSALOauth2ProviderFactory.m]: We have an error for unsupported feature and we can bubble that up as a response, no?

File: MSAL/src/instance/MSALOauth2ProviderFactory.m [hunk 2]
```diff
@@ -25,34 +25,44 @@
...
+        return nil;
+    }
+    
+    // Create base factory for everything else, but in future we might want to further separate this out
     // (e.g. ADFS, Google, Oauth2 etc...)
-    return [MSIDAADV2Oauth2Factory new];
+    return [MSALOauth2Provider new];
```
Comment (unpluggedk) [inline MSAL/src/instance/MSALOauth2ProviderFactory.m]: For now, this really shouldn't be reached, should we bubble up parameter error?

File: MSAL/src/instance/oauth2/MSALOauth2Provider+Internal.h [hunk 3]
```diff
@@ -25,14 +25,20 @@
...
+#define MSALOauth2BaseFactory_Internal_h
 
-+ (nullable MSIDOauth2Factory *)msidOauth2FactoryForAuthority:(nonnull NSURL *)authority
-                                                      context:(nullable id<MSIDRequestContext>)context
-                                                        error:(NSError * _Nullable __autoreleasing * _Nullable)error;
+@interface MSALOauth2Provider()
+
+@property (nonatomic, nonnull, readwrite) MSIDOauth2Factory *msidOauth2Factory;
+
+- (void)initDerivedProperties;
```
Comment (unpluggedk) [inline MSAL/src/instance/oauth2/MSALOauth2Provider+Internal.h]: nit: should we be more specific? and does this need to be public? i.e/ would other class call this method directly?

File: MSAL/src/instance/oauth2/MSALOauth2Provider.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+@class MSIDTokenResult;
+@class MSIDDefaultTokenCacheAccessor;
+@class MSALAccount;
+@class MSIDAuthority;
+
+@interface MSALOauth2Provider : NSObject
+
+@property (nonatomic, readonly) MSIDOauth2Factory *msidOauth2Factory;
+
+- (nullable MSALResult *)resultWithTokenResult:(nonnull MSIDTokenResult *)tokenResult
```
Comment (unpluggedk) [inline MSAL/src/instance/oauth2/MSALOauth2Provider.h]: nit: it is already bound by nonnull macro..

Comment (oldalton) [reply MSAL/src/instance/oauth2/MSALOauth2Provider.h]: Fixed

File: MSAL/src/instance/oauth2/MSALOauth2Provider+Internal.h (hunk 3 above)
Comment (oldalton) [reply MSAL/src/instance/oauth2/MSALOauth2Provider+Internal.h]: Fixed

File: MSAL/src/instance/MSALOauth2ProviderFactory.m (hunk 2 above)
Comment (oldalton) [reply MSAL/src/instance/MSALOauth2ProviderFactory.m]: Fixed

File: MSAL/src/instance/MSALOauth2ProviderFactory.m (hunk 1 above)
Comment (oldalton) [reply MSAL/src/instance/MSALOauth2ProviderFactory.m]: Fixed

**Pull request merged**